File: scripts/libraries/light/lightSource.js
//light source, does all the calculations at placement and when any block has been placed.
//calc method: first raytrace, then if updated check if block placement area(x, y) intersects with each straight line from the source(aka the theta is the same).

const clone = obj => {
	if(obj === null || typeof(obj) !== 'object')
	return obj;

	var copy = obj.constructor();

	for(var attr in obj){
		if(obj.hasOwnProperty(attr)){
			copy[attr] = obj[attr];
		}
	};

	return copy;
}

module.exports = {
	extend(type, build, name, obj, objb){
		if(obj == undefined) obj = {};
		if(objb == undefined) objb = {};
		obj = Object.assign({
			//start
      //The strength of light, used by consumers
			lightStrength: 60,
      //The distance light does before dissapating
      lightLength: 50,
      //The absolute distance a light can reach, regardless of lightRepeaters
      maxLightLength: 500,
      //the max reflections this light has, everything that affects this light is consided a reflection
      maxReflections: 50,
      lightColor: Color.white,
      //whether to scale lightStrength with the input power status
      scaleStatus: true,
      //whether to display angle configuration
      angleConfig: false,
      //defaults
      dirs: [[1,0], [1,1], [0,1], [-1,1], [-1,0], [-1,-1], [0,-1], [1,-1]],
      update: true,
      rotate: true
			//end
		}, obj, {
			//start
			setBars(){
				this.super$setBars();

				this.bars.add("light", func(build => {
					return new Bar(
						prov(() => Core.bundle.format("lightlib.light", build.lightPower())),
						prov(() => this.lightColor),
						floatp(() => {
							return build.lightPower() / this.lightStrength;
						})
					);
				}));
			}
			//end
		});
		print("Created Block: " + Object.keys(obj));
		const lightblock = extendContent(type, name, obj);
		lightblock.configurable = lightblock.angleConfig;
    lightblock.reflowTimer = lightblock.timers++;
    if(lightblock.angleConfig) lightblock.rotate = false;
    if(lightblock.size%2 == 0) print("[scarlet]Even - sized light blocks are not supported! Continue anyways?[]");

		//lightblock.hasLevelFunction = (typeof objb["levelUp"] === "function");
		lightblock.hasCustomUpdate = (typeof objb["customUpdate"] === "function");
    lightblock.hasCustomRW = (typeof objb["customRead"] === "function");

		objb = Object.assign(objb, {
      //angle strengthPercentage lengthleft color
      _lightData: [0, 100, lightblock.lightLength, lightblock.lightColor],
      //array of tiles
      _ls: [],
      //array of _lightData for each _ls
      _lsData: [],
      _lCons: [],
      //only used in angleConfig
      _angle: 0,
      _strength: 0,
      _lightInit: false,

      setInit(a){
        this._lightInit = a;
      },
      initDone(){
        return this._lightInit;
      },

      getAngleDeg(){
        return (lightblock.rotate)?this.rotDeg():this._angle*45;
      },
      getAngle(){
        return (lightblock.rotate)?this.rotation*2:this._angle;
      },
      setAngle(a){
        this._angle = a;
        this.lightMarchStart(lightblock.lightLength, lightblock.maxLightLength);
      },

      lightData(){
        return this._lightData;
      },
      setLightData(d){
        this._lightData = d;
      },
      setPower(a){
        this._strength = a;
      },

      lightPower(){
        if(!this._lightInit) return this.targetStrength();
        return this._strength;
      },

      targetStrength(){
        if(!this.cons.valid()) return 0;
        return (lightblock.scaleStatus)?lightblock.lightStrength * this.power.status:lightblock.lightStrength;
      },

			updateTile(){
        this.setPower(this.targetStrength());
        if(!this.initDone()) this.lightMarchStart(lightblock.lightLength, lightblock.maxLightLength);
        else if(this.timer.get(lightblock.reflowTimer, 30) && this.lightPower() > 1) this.lightMarchStart(lightblock.lightLength, lightblock.maxLightLength);

				if(lightblock.hasCustomUpdate) this.customUpdate();
				else this.super$updateTile();
			},

			read(stream, version){
				this.super$read(stream, version);
				this._angle = stream.b();
        if(lightblock.hasCustomRW) this.customRead(stream, version);
			},
			write(stream){
				this.super$write(stream);
				stream.b(this._angle);
        if(lightblock.hasCustomRW) this.customWrite(stream);
			},

      drawLight(){
        //TODO make light draw on beams
        Drawf.light(this.team, this.x, this.y, (this.lightPower()*0.1 + 60)*this.power.status, lightblock.lightColor, 0.8);
      },

      drawLightLasers(){
        if(this == null || !this.isAdded() || this.lightPower() <= 1) return;
        Draw.z(Layer.effect - 1);
        Draw.blend(Blending.additive);

        //var now = null;
        //var next = null;
        for(var i=0; i<this._ls.length; i++){
          if(this._lsData[i] == null) continue;
          //print("Drawing Data: "+this._lsData[i]);
          var a = this._lsData[i][1]/100*(this.lightPower()/lightblock.lightStrength);
          Draw.color(this._lsData[i][3], a);
          if(Core.settings.getBool("bloom")) Draw.z((a>0.99)?(Layer.effect - 1):(Layer.bullet - 2));
          Lines.stroke(1 + Math.min(this.lightPower()/1000, 10));
          //now = this._ls[i];
          //next = this._ls[i+1];
          if(i == this._ls.length - 1 || this._ls[i+1] == null){
            //I'm sorry. okay?
            Draw.alpha(a);
            Lines.lineAngle(this._ls[i].worldx(), this._ls[i].worldy(), this._lsData[i][0]*45, this._lsData[i][2]*Vars.tilesize);
            Draw.alpha(a*0.5);
            Lines.lineAngle(this._ls[i].worldx(), this._ls[i].worldy(), this._lsData[i][2]*Vars.tilesize, this._lsData[i][0]*45, 4);
            Draw.alpha(a*0.25);
            Lines.lineAngle(this._ls[i].worldx(), this._ls[i].worldy(), this._lsData[i][2]*Vars.tilesize + 4, this._lsData[i][0]*45, 2);
            Draw.alpha(a*0.125);
            Lines.lineAngle(this._ls[i].worldx(), this._ls[i].worldy(), this._lsData[i][2]*Vars.tilesize + 6, this._lsData[i][0]*45, 2);
          }
          else{
            if(this._lsData[i+1] == null){
              //obstructed
              Lines.line(this._ls[i].worldx(), this._ls[i].worldy(), this._ls[i+1].worldx() - 4 * lightblock.dirs[this._lsData[i][0]][0], this._ls[i+1].worldy() - 4 * lightblock.dirs[this._lsData[i][0]][1]);
            }
            else{
              //light go brrrrrrrr
              Lines.line(this._ls[i].worldx(), this._ls[i].worldy(), this._ls[i+1].worldx(), this._ls[i+1].worldy());
            }
          }
        }
        Draw.blend();
        Draw.color();
      },

      lightMarchStart(length, maxLength){
        //idk
        this._lightData[0] = this.getAngle();
        this._lightData[1] = 100;
        //TODO make it more efficient
        for(var i=0; i<this._lCons.length; i++){
          this._lCons[i].removeSource(this);
        }
        this._ls = [];
        this._lsData = [];
        this._lCons = [];
        this._ls.push(this.tile);
        this._lsData.push([this.getAngle(), 100, this._lightData[2], this._lightData[3]]);
        this.pointMarch(this.tile, this.lightData(), length, maxLength, 0, this);
        this.setInit(true);
        //print(this._ls.toString());
        //print(this._lsData.toString());
      },
      pointMarch(tile, ld, length, maxLength, num, source){
        if(length <= 0 || maxLength <= 0 || num > lightblock.maxReflections || ld[1]*source.lightPower() < 1) return;
        const dir = lightblock.dirs[ld[0]];
        var next = null;
        var next2 = null;
        var furthest = null;
        var i = 0;
        var hit = Vars.world.raycast(tile.x, tile.y, tile.x + length*dir[0], tile.y + length*dir[1], (x, y)=>{
          furthest = Vars.world.tile(x, y);
          if(furthest == tile || furthest == null) return false;
          i++;
          if(!furthest.solid()) return false;
          if(furthest.bc() == null) return true;
          if(furthest.bc().block.lightReflector){
            //print("Light reflector!");
            var tr = furthest.bc().calcReflection(ld[0]);
            if(tr >= 0) next = [tr, ld[1], ld[2] - i, ld[3]];
          }
          else if(furthest.bc().block.lightDivisor){
            var tr = furthest.bc().calcReflection(ld[0]);
            if(tr >= 0){
              next = [ld[0], ld[1] / 2, ld[2] - i, ld[3]];
              next2 = [tr, ld[1] / 2, ld[2] - i, ld[3]];
            }
          }
          else if(furthest.bc().block.lightRepeater){
            var tl = furthest.bc().calcLight(ld, i);
            if(tl == null) return true;
            next = [tl[0], tl[1], tl[2], tl[3]];
          }
          else if(furthest.bc().block.consumesLight){
            furthest.bc().addSource([source, ld[1]]);
            this._lCons.push(furthest.bc());
          }
          return true;
        });
        if(!hit) return;
        if(next == null){
          //the block hit is solid or a consumer
          this._ls.push(furthest);
          this._lsData.push(null);//obstructor
        }
        else if(next2 == null){
          //the block hit reflecc
          this._ls.push(furthest);
          this._lsData.push(next);//mirror
          this.pointMarch(furthest, next, ld[2]-i, maxLength-i, ++num, source);
        }
        else{
          //the light go S P L I T
          //TODO
          this._ls.push(furthest);
          this._lsData.push(next);//mirror
          this.pointMarch(furthest, next, ld[2]-i, maxLength-i, ++num, source);
          this._ls.push(null);//cheaty yep
          this._lsData.push(null);
          this._ls.push(furthest);
          this._lsData.push(next2);//mirror mirror on the wall
          this.pointMarch(furthest, next2, ld[2]-i, maxLength-i, ++num, source);
        }
      }
		});
		//Extend Building
		print("Prep Building: " + Object.keys(objb));
		lightblock.entityType = ent => {
			ent = extendContent(build, lightblock, clone(objb));
			ent._angle = 0;
      ent._lightInit = false;
      Events.run(Trigger.draw, () => {
        if(ent != null) ent.drawLightLasers();
      });
			return ent;

		};

		return lightblock;
	}
}

File: main/src/unity/world/blocks/logic/LightSource.java
package unity.world.blocks.logic;

import java.util.ArrayList;
import arc.Events;
import arc.graphics.*;
import arc.graphics.g2d.*;
import arc.util.Eachable;
import arc.util.io.*;
import arc.math.geom.*;
import arc.scene.ui.layout.Table;
import mindustry.gen.*;
import mindustry.entities.units.BuildPlan;
import mindustry.game.EventType.Trigger;
import mindustry.graphics.*;
import mindustry.ui.*;
import mindustry.world.Tile;
import mindustry.world.meta.*;
import mindustry.world.blocks.production.GenericCrafter;
import unity.world.blocks.LightData;
import unity.world.blocks.LightRepeaterBuildBase;
import unity.world.blocks.logic.LightGenerator.LightGeneratorBuild;
import unity.world.blocks.logic.LightReflector.LightReflectorBuild;

import static arc.Core.*;
import static mindustry.Vars.*;

public class LightSource extends GenericCrafter{
    public int lightStrength = 60, lightLength = 50, maxLightLength = 5000, maxReflections = 128, lightInterval = 20;
    protected final int reflowTimer = timers++;
    protected Color lightColor = Color.white;
    protected boolean scaleStatus = true;
    public final boolean hasCustomUpdate, angleConfig;
    public TextureRegion baseRegion, topRegion, lightRegion, liquidRegion;

    public LightSource(String name, boolean hasCustomUpdate, boolean angleConfig){
        super(name);
        this.hasCustomUpdate = hasCustomUpdate;
        this.angleConfig = angleConfig;
        update = true;
        rotate = true;
        if(angleConfig){
            rotate = false;
            configurable = true;
            saveConfig = true;
            lastConfig = 0;
            config(Integer.class, (LightSourceBuild build, Integer value) -> {
                build.addAngle(value);
            });
        }
    }

    public LightSource(String name, boolean angleConfig){
        this(name, false, angleConfig);
    }

    public LightSource(String name){
        this(name, false);
    }

    @Override
    public void setBars(){
        super.setBars();
        bars.add("light", (LightSourceBuild build) -> new Bar(() -> bundle.format("lightlib.light", build.getStrength()),
        () -> lightColor, () -> build.getStrength() / lightStrength));
    }

    @Override
    public void setStats(){
        super.setStats();
        stats.add(Stat.output, "@ @", bundle.format("lightlib.light", lightStrength), StatUnit.perSecond.localized());
    }

    @Override
    public void drawRequestRegion(BuildPlan req, Eachable<BuildPlan> list){
        final float scl = tilesize * req.animScale * size;
        Draw.rect(baseRegion, req.drawx(), req.drawy(), scl, scl);
        if(configurable){
            if(req.config != null) drawRequestConfig(req, list);
        }else Draw.rect(topRegion, req.drawx(), req.drawy(), scl, scl, req.rotation * 90f);
    }

    @Override
    public void drawRequestConfig(BuildPlan req, Eachable<BuildPlan> list){
        final float scl = tilesize * req.animScale * size;
        Draw.rect(topRegion, req.drawx(), req.drawy(), scl, scl, req.config == null ? 0 : ((int) req.config) * 45f);
    }

    public class LightSourceBuild extends GenericCrafterBuild{
        protected Tile furthest;
        protected float strength = 0f;
        protected boolean lightInit = false;
        protected int angle = 0, loops;
        protected final ArrayList<Tile> ls = new ArrayList<>();
        protected final ArrayList<LightData> lsData = new ArrayList<>();
        protected final ArrayList<LightGeneratorBuild> lCons = new ArrayList<>();
        protected LightData lightData = new LightData(lightLength, lightColor);

        {
            Events.run(Trigger.draw, () -> {
                if(this != null) drawLightLasers();
            });
        }

        protected void setInit(boolean a){
            lightInit = a;
        }

        protected boolean initDone(){
            return lightInit;
        }

        public float getAngleDeg(){
            return rotate ? rotdeg() : angle * 45f;
        }

        protected int getAngle(){
            return rotate ? rotation * 2 : angle;
        }

        protected void setAngle(int a){
            angle = a;
            lightMarchStart(lightLength, maxLightLength);
        }

        public void addAngle(int a){
            setAngle((angle + a + 8) % 8);
        }

        protected void setStrength(float a){
            strength = a;
        }

        public float getStrength(){
            if(!lightInit) return targetStrength();
            return strength;
        }

        protected float getPowerStatus(){
            if(!hasPower || power == null) return 1f;
            return power.status;
        }

        protected float targetStrength(){
            if(!cons.valid()) return 0f;
            return scaleStatus ? lightStrength * getPowerStatus() : lightStrength;
        }

        @Override
        public void updateTile(){
            setStrength(targetStrength());
            if(!initDone()) lightMarchStart(lightLength, maxLightLength);
            else if(lightInterval <= 0 || timer.get(reflowTimer, lightInterval)){
                if(getStrength() > 1f) lightMarchStart(lightLength, maxLightLength);
                else clearCons();
            }
            if(hasCustomUpdate) lightSourceUpdate();
            else super.updateTile();
        }

        protected void lightSourceUpdate(){

        }

        @Override
        public void write(Writes write){
            super.write(write);
            write.b(angle);
            lightSourceWrite(write);
        }

        protected void lightSourceWrite(Writes write){

        }

        @Override
        public void read(Reads read, byte revision){
            super.read(read);
            angle = read.b();
            lightSourceRead(read, revision);
        }

        protected void lightSourceRead(Reads read, byte revision){

        }

        @Override
        public void drawLight(){
            Drawf.light(team, x, y, (getStrength() * 0.1f + 60f) * getPowerStatus(), lightColor, 0.8f);
        }

        protected void drawLightLasers(){
            if(!isAdded() || getStrength() <= 1) return;
            Draw.z(Layer.effect - 1f);
            Draw.blend(Blending.additive);

            final float w = 1 + Math.min(getStrength() / 1000f, 10f);
            Lines.stroke(w);
            for(int i = 0, len = ls.size(); i < len; i++){
                LightData tempData = lsData.get(i);
                if(tempData == null) continue;
                Tile tempTile = ls.get(i);

                int tempAngle = tempData.angle;
                float tempLength = tempData.length, tempX = tempTile.worldx(), tempY = tempTile.worldy();
                float a = tempData.strength / 100f * (getStrength() / lightStrength);

                Draw.color(tempData.color, a);
                if(settings.getBool("bloom")) Draw.z(a > 0.99f ? Layer.effect - 1 : Layer.bullet - 2);
                if(i == len - 1 || ls.get(i + 1) == null){
                    if(tempData.length < 1) continue;
                    Lines.line(tempX, tempY, tempX + (tempLength - 1) * tilesize * Geometry.d8(tempAngle).x, tempY + (tempLength - 1) * tilesize * Geometry.d8(tempAngle).y);
                    final float shift = (tempAngle % 2 == 0) ? w / 2f : w / 2.8285f;
                    Drawf.tri(tempX + ((tempLength - 1) * tilesize + shift) * Geometry.d8(tempAngle).x, tempY + ((tempLength - 1) * tilesize + shift) * Geometry.d8(tempAngle).y, w, tempAngle % 2 == 0 ? 8f : 11.313f, tempData.angle * 45f);
                }else{
                    Tile nextTile = ls.get(i + 1);
                    if(lsData.get(i + 1) == null) Lines.line(tempX, tempY, nextTile.worldx() - 4 * Geometry.d8(tempAngle).x, nextTile.worldy() - 4 * Geometry.d8(tempAngle).y);
                    else Lines.line(tempX, tempY, nextTile.worldx(), nextTile.worldy());
                }
            }
            Draw.blend();
            Draw.color();
        }

        protected void clearCons(){
            lCons.forEach(b -> b.removeSource(this));
        }

        protected void lightMarchStart(int length, int maxLength){
            lightData.angle = getAngle();
            lightData.strength = 100f;
            clearCons();
            ls.clear();
            lsData.clear();
            lCons.clear();
            ls.add(tile);
            lsData.add(lightData);
            pointMarch(tile, lightData, length, maxLength, 0);
            setInit(true);
        }

        protected void pointMarch(Tile tile, LightData ld, int length, int maxLength, int num){
            if(length <= 0 || maxLength <= 0 || ld.strength * getStrength() < 1f) return;

            final Point2 dir = Geometry.d8(ld.angle);
            LightData next = new LightData();
            LightData next2 = new LightData();
            furthest = null;
            loops = 0;

            boolean hit = world.raycast(tile.x, tile.y, tile.x + length * dir.x, tile.y + length * dir.y, (x, y) -> {
                furthest = world.tile(x, y);
                if(furthest == tile || furthest == null) return false;
                loops++;

                if(!furthest.solid() || (furthest.block() == block && tile == this.tile)) return false;
                Building build = furthest.bc();
                if(build == null) return true;
                if(build.block instanceof LightDivisor){
                    int tr = ((LightReflectorBuild) build).calcReflection(ld.angle);
                    if(tr >= 0){
                        next.set(ld.angle, ld.strength / 2f, ld.length - loops, ld.color);
                        next2.set(tr, ld.strength / 2f, ld.length - loops, ld.color);
                    }
                }else if(build instanceof LightReflectorBuild){
                    int tr = ((LightReflectorBuild) build).calcReflection(ld.angle);
                    if(tr >= 0) next.set(tr, ld.strength, ld.length - loops, ld.color);
                }else if(build instanceof LightRepeaterBuildBase){
                    LightData tempData = ((LightRepeaterBuildBase) build).calcLight(ld, loops);
                    if(tempData == null) return true;
                    next.set(tempData);
                }else if(build.block instanceof LightGenerator)
                    lCons.add(((LightGeneratorBuild) build).addSource(this, ld));
                return true;
            });

            if(!hit) return;
            if(!next.isIntialized() || num > maxReflections){
                ls.add(furthest);
                lsData.add(null);
            }else if(!next2.isIntialized()){
                ls.add(furthest);
                lsData.add(next);
                pointMarch(furthest, next, ld.length - loops, maxLength - loops, ++num);
            }else{
                ls.add(furthest);
                lsData.add(next);
                Tile forSafety = furthest;
                pointMarch(furthest, next, ld.length - loops, maxLength - loops, ++num);
                ls.add(null);
                lsData.add(null);
                ls.add(forSafety);
                lsData.add(next2);
                pointMarch(forSafety, next2, ld.length - loops, maxLength - loops, ++num);
            }
        }

        @Override
        public Integer config(){
            return angle;
        }

        @Override
        public void buildConfiguration(Table table){
            table.button(Icon.leftOpen, Styles.clearTransi, 34f, () -> {
                configure(1);
            }).size(40f);
            table.button(Icon.rightOpen, Styles.clearTransi, 34f, () -> {
                configure(1);
            }).size(40f);
        }
    }
}
